proto/structured.go

```diff
@@ -0,0 +1,185 @@
+		}
+		columnIDs[column.ID] = column.Name
+
+		if column.ID >= desc.NextColumnID {
+			return fmt.Errorf("column \"%s\" invalid ID (%d) > next column ID (%d)",
+				column.Name, column.ID, desc.NextColumnID)
+		}
+	}
+
+	// TODO(pmattis): Check that the indexes are unique. That is, no 2 indexes
+	// should contain identical sets of columns.
+
+	if len(desc.Indexes) == 0 {
+		return fmt.Errorf("table must contain at least 1 index")
+	}
+
+	indexNames := map[string]struct{}{}
+	indexIDs := map[uint32]string{}
+	for _, index := range desc.Indexes {
+		if err := validateName(index.Name, "index"); err != nil {
+			return err
+		}
+
+		if _, ok := indexNames[index.Name]; ok {
+			return fmt.Errorf("duplicate index name: \"%s\"", index.Name)
+		}
+		indexNames[index.Name] = struct{}{}
+
+		if other, ok := indexIDs[index.ID]; ok {
+			return fmt.Errorf("index \"%s\" duplicate ID of index \"%s\": %d",
+				index.Name, other, index.ID)
+		}
+		indexIDs[index.ID] = index.Name
+
+		if index.ID >= desc.NextIndexID {
+			return fmt.Errorf("index \"%s\" invalid index ID (%d) > next index ID (%d)",
+				index.Name, index.ID, desc.NextIndexID)
+		}
+
+		if len(index.ColumnIDs) == 0 {
+			return fmt.Errorf("index \"%s\" must contain at least 1 column", index.Name)
+		}
+
+		for _, columnID := range index.ColumnIDs {
+			if _, ok := columnIDs[columnID]; !ok {
+				return fmt.Errorf("index \"%s\" contains unknown column ID %d", index.Name, columnID)
+			}
+		}
+	}
+	return nil
+}
+
+// TableDescFromSchema initializes a TableDescriptor from a TableSchema. The
+// TableSchema is expected to be valid. An invalid table schema will result in
+// an invalid table descriptor. Call ValidateTableDesc on the resulting
+// descriptor to check for validity.
+//
+// Note that the resulting descriptor will not have a table ID set. Allocation
+// of the table ID is left to the caller.
+func TableDescFromSchema(schema TableSchema) TableDescriptor {
+	desc := TableDescriptor{
+		Table: schema.Table,
+	}
+	desc.Name = strings.ToLower(desc.Name)
+
+	columnIDsByName := map[string]uint32{}
+	for _, column := range schema.Columns {
+		columnDesc := ColumnDescriptor{
+			ID:     desc.NextColumnID,
+			Column: column,
+		}
+		columnDesc.Name = strings.ToLower(columnDesc.Name)
+		columnIDsByName[columnDesc.Name] = columnDesc.ID
+
+		desc.Columns = append(desc.Columns, columnDesc)
+		desc.NextColumnID++
+	}
+
+	for _, index := range schema.Indexes {
+		indexDesc := IndexDescriptor{
+			ID:    desc.NextIndexID,
+			Index: index.Index,
+		}
+		indexDesc.Name = strings.ToLower(indexDesc.Name)
+
+		for _, columnName := range index.ColumnNames {
+			indexDesc.ColumnIDs = append(indexDesc.ColumnIDs, columnIDsByName[columnName])
+		}
+
+		desc.Indexes = append(desc.Indexes, indexDesc)
+		desc.NextIndexID++
+	}
+
+	return desc
+}
+
+// TableSchemaFromDesc initializes a TableSchema from a TableDescriptor. The
+// TableDescriptor is expected to be valid. An invalid table descriptor will
```

Comment:
Hey since we are reading the descriptor from disk and it can be changed through a backdoor, I think we should do some amount of validation of descriptors.